src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractJenkinsMojo.java

```diff
@@ -89,6 +94,39 @@ protected String findJenkinsVersion() throws MojoExecutionException {
         throw new MojoExecutionException("Failed to determine Jenkins version this plugin depends on.");
     }
 
+    protected JavaSpecificationVersion getMinimumJavaVersion() throws MojoExecutionException {
+        Artifact bom = resolveJenkinsCoreBom(findJenkinsVersion());
+
+        Properties properties;
+        try {
+            properties = wrap(bom).resolvePom().getProperties();
+        } catch (ProjectBuildingException e) {
+            throw new MojoExecutionException("Failed to resolve artifact " + bom, e);
+        }
+
+        String version = properties.getProperty("java.level");
+        if (version == null || version.isEmpty()) {
+            throw new MojoExecutionException("java.level not defined in " + bom);
+        }
+        return new JavaSpecificationVersion(version);
+    }
+
+    private Artifact resolveJenkinsCoreBom(String version) throws MojoExecutionException {
+        DefaultArtifactCoordinate artifactCoordinate = new DefaultArtifactCoordinate();
+        artifactCoordinate.setGroupId("org.jenkins-ci.main");
+        artifactCoordinate.setArtifactId("jenkins-bom");
+        artifactCoordinate.setVersion(version);
+        artifactCoordinate.setExtension("pom");
+
+        try {
+            return artifactResolver
+                    .resolveArtifact(session.getProjectBuildingRequest(), artifactCoordinate)
+                    .getArtifact();
+        } catch (ArtifactResolverException e) {
+            throw new MojoExecutionException("Couldn't download artifact: ", e);
```

Comment:
IIUC this will be the failure point if your `jenkins.version` predates the version where the core BOM was introduced? So it would be polite to make the error message just say what that minimum version is.

Reply:
Ah I see your point. Well the next check takes care of that by requiring 2.204 or later. So I can just reverse the order of the checks to make the message more polite.